qrest/exception.py

```diff
@@ -78,62 +77,71 @@ class RestResourceMissingContentError(RestClientResourceError):
     pass
 
 
-class RestResourceNotFoundError(RestClientResourceError):
+class RestCredentailsError(RestClientResourceError):
     """ wrapper exception """
 
     pass
 
 
-class RestAccessDeniedError(RestClientResourceError):
-    """ wrapper exception """
+class RestResponseError(RestClientResourceError):
+    """Exception due to a requests.Response whose status code indicates an error.
 
-    pass
+    :ivar response: Response object returned by the request
 
+    """
 
-class RestCredentailsError(RestClientResourceError):
-    """ wrapper exception """
+    response: Response
 
-    pass
+    def __init__(self, response: Response, message: str):
+        super().__init__(message)
+        self.response = response
 
 
-class RestInternalServerError(RestClientResourceError):
-    """ wrapper exception """
+class RestAccessDeniedError(RestResponseError):
+    def __init__(self, response: Response):
+        super().__init__(
+            response, f"error {response.status_code}: Access is denied to resource {response.url}"
+        )
 
-    pass
 
+class RestBadRequestError(RestResponseError):
+    def __init__(self, response: Response):
+        super().__init__(response, f"Bad request for resource {response.url}")
```

Comment:
Same comment as with the `RestResourceNotFoundError`. For a 400 response, it is useful for the user to know what the exact error is.